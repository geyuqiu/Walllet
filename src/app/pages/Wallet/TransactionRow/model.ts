export interface Transaction {
	id: string;
	amount: string;
	fee: string;
	sender: string;
	recipient: string;
	timestamp: string;
};
