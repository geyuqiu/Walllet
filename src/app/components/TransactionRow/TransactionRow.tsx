import React from "react";

import {Icon} from '../Icon';
import {TableCell} from '../Table/TableCell';
import {TableRow} from "../Table/TableRow";
import {Transaction} from './model';

type TransactionRowProps = {
	transaction: Transaction;
	address: string
};

export const TransactionRow = ({transaction, address}: TransactionRowProps) => {
	const trimLongText = (id: string) => `${id.substring(0, 5)}...${id.substring(id.length - 5, id.length)}`
	return (
		<TableRow>
			<TableCell innerClassName="justify-start"
			           className="sm:text-sm sm:p-3 sm:first:pl-0 sm:last:pr-0">
				<a href={'https://explorer.ark.io/transaction/' + transaction.id}
				   className="text-green-dark">
					<span className="table-cell lg:table-cell sm:hidden">{trimLongText(transaction.id)}</span>
					<Icon name="Txid" width={20} height={20} className="hidden lg:hidden sm:table-cell"/>
				</a>
			</TableCell>
			<TableCell innerClassName="justify-start border-l border-theme-secondary-300">
				<a href={'https://explorer.ark.io/wallets/' + transaction.sender}
					 className={address === transaction.sender ? "text-green-dark" : ""}
				>{trimLongText(transaction.sender)}</a>
			</TableCell>
			<TableCell innerClassName="justify-start border-l border-theme-secondary-300">
				<a href={'https://explorer.ark.io/wallets/' + transaction.recipient}
				   className={address === transaction.recipient ? "text-green-dark" : ""}
				>{trimLongText(transaction.recipient)}</a>
			</TableCell>
			<TableCell innerClassName="justify-center border-l border-theme-secondary-300"
			           className="hidden xl:table-cell"
			>
				{transaction.timestamp}
			</TableCell>
			<TableCell innerClassName="justify-end border-l border-theme-secondary-300">
				{transaction.amount}
			</TableCell>
			<TableCell innerClassName="justify-end border-l border-theme-secondary-300"
			           className="hidden lg:table-cell"
			>
				{transaction.fee}
			</TableCell>
		</TableRow>
	);
}
