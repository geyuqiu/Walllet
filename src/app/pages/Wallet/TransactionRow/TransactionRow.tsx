import {BigNumber} from '@arkecosystem/platform-sdk-support';
import {Icon} from "app/components/Icon/Icon";
import React from "react";

import {Amount} from '../../../components/Amount/Amount';
import {Link} from '../../../components/Link/Link';
import {TableCell} from '../../../components/Table/TableCell';
import {TableRow} from "../../../components/Table/TableRow";
import {Transaction} from './model';


type TransactionRowProps = {
	transaction: Transaction;
	address: string;
};

export const hideTextBetween = ({id, prefixLength, suffixLength}: { id?: string, prefixLength: number, suffixLength: number }) =>
	id ? `${id.substring(0, prefixLength)}...${id.substring(id.length - suffixLength, id.length)}` : '';

export const TransactionRow = ({transaction, address}: TransactionRowProps) => {
	const linkClassName = "text-green-dark hover:text-green-darkest active:text-green-light font-semibold hover:underline";
	return (
		<TableRow className="text-base">

			<TableCell innerClassName="justify-end md:justify-start"
			           className="md:py-2 md:w-4">
				<Link href={'https://explorer.ark.io/transaction/' + transaction.id}
				      className={linkClassName}>
					<span className="table-cell lg:table-cell md:hidden">{hideTextBetween({
						id: transaction.id,
						prefixLength: 5,
						suffixLength: 5
					})}</span>
					<Icon name="Txid" width={20} height={20} className="hidden lg:hidden md:table-cell"/>
				</Link>
			</TableCell>

			<TableCell innerClassName="justify-end md:justify-start md:border-l border-theme-secondary-300">
				{address !== transaction.sender
					? <Link href={'https://explorer.ark.io/wallets/' + transaction.sender} className={linkClassName}>
						<ResponsiveText text={transaction.sender} prefixLength={5} suffixLength={5}/>
					</Link>
					: <ResponsiveText text={transaction.sender} prefixLength={5} suffixLength={6}/>
				}
			</TableCell>

			<TableCell innerClassName="justify-end md:justify-start md:border-l border-theme-secondary-300">
				{address !== transaction.recipient
					? <Link href={'https://explorer.ark.io/wallets/' + transaction.recipient} className={linkClassName}>
						<ResponsiveText text={transaction.recipient} prefixLength={5} suffixLength={5}/>
					</Link>
					: <ResponsiveText text={transaction.recipient} prefixLength={5} suffixLength={6}/>
				}
			</TableCell>

			<TableCell innerClassName="justify-end md:justify-center md:border-l border-theme-secondary-300"
			           className="hidden xl:table-cell">
				{transaction.timestamp}
			</TableCell>

			<TableCell innerClassName="justify-end md:border-l border-theme-secondary-300">
				<Amount
					data-testid="transaction__fee__Amount"
					value={BigNumber.make(transaction.amount)}
				/>
			</TableCell>

			<TableCell innerClassName="justify-end md:border-l border-theme-secondary-300"
			           className="hidden lg:table-cell">
				<Amount
					data-testid="transaction__fee__Amount"
					value={BigNumber.make(transaction.fee)}
				/>
			</TableCell>

		</TableRow>
	);
}

type ResponsiveTextProps = {
	text: string;
	prefixLength: number;
	suffixLength: number;
};

export const ResponsiveText = ({text, prefixLength, suffixLength}: ResponsiveTextProps) =>
	<>
		<span className="hidden sm:block md:hidden">{text}</span>
		<span className="block sm:hidden md:block">{hideTextBetween({
			id: text,
			prefixLength: prefixLength,
			suffixLength: suffixLength
		})}</span>
	</>
