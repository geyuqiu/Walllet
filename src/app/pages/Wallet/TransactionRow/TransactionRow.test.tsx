import {render} from "@testing-library/react";
import React from "react";

import {Transaction} from './model';
import {hideTextBetween, TransactionRow} from "./TransactionRow";

describe("Footer", () => {
	it("should match snapshot", () => {
		const transaction = {
			id: "77a740c7517ae164b35e4e28da7251543d9d98e34e90d3baf604fc582159c352",
			amount: "100000000000000",
			fee: "1000000",
			sender: "AReY3W6nTv3utiG2em5nefKEsGQeqEVPN4",
			recipient: "AdzbhuDTyhnfAqepZzVcVsgd1Ym6FgETuW",
			timestamp: "2021-04-01T11:41:20.000Z"
		} as Transaction;

		const { container } = render(
				<TransactionRow transaction={transaction} address={''}/>
		);

		expect(container).toMatchSnapshot();
	});
});

describe("trimLongText", () => {
	it("return 5 characters as prefix and 6 characters as suffx", () => {
		const expected = hideTextBetween('AdzbhuDTyhnfAqepZzVcVsgd1Ym6FgETuW', 5, 6);

		expect(expected).toEqual('Adzbh...FgETuW');
	});
	it("return same number of characters as prefix and as suffx", () => {
		const expected = hideTextBetween('AdzbhuDTyhnfAqepZzVcVsgd1Ym6FgETuW', 5, 5);

		expect(expected).toEqual('Adzbh...gETuW');
	});
});
