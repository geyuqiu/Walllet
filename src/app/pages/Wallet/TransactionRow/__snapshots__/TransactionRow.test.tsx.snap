// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer should match snapshot 1`] = `
<div>
  <tr
    class="TableRow__TableRowStyled-sc-1p89o19-0 iIKWLH"
    data-testid="TableRow"
  >
    <td
      class="sm:pt-3 sm:pb-3"
    >
      <div
        class="TableCell__TableCellInnerWrapper-sc-12y9mmp-0 dbZfon justify-end sm:justify-start"
      >
        <a
          class="text-green-dark hover:text-green-darkest active:text-green-light font-semibold hover:underline"
          href="https://explorer.ark.io/transaction/77a740c7517ae164b35e4e28da7251543d9d98e34e90d3baf604fc582159c352"
        >
          <span
            class="table-cell lg:table-cell sm:hidden"
          >
            77a74...9c352
          </span>
          <div
            class="sc-bdfBwQ iyevwF hidden lg:hidden sm:table-cell"
            height="20"
            width="20"
          >
            <svg>
              txid.svg
            </svg>
          </div>
        </a>
      </div>
    </td>
    <td>
      <div
        class="TableCell__TableCellInnerWrapper-sc-12y9mmp-0 dbZfon justify-end sm:justify-start sm:border-l border-theme-secondary-300 pl-5 pr-5"
      >
        AReY3...EVPN4
      </div>
    </td>
    <td>
      <div
        class="TableCell__TableCellInnerWrapper-sc-12y9mmp-0 dbZfon justify-end sm:justify-start sm:border-l border-theme-secondary-300"
      >
        Adzbh...gETuW
      </div>
    </td>
    <td
      class="hidden xl:table-cell"
    >
      <div
        class="TableCell__TableCellInnerWrapper-sc-12y9mmp-0 dbZfon justify-end sm:justify-center sm:border-l border-theme-secondary-300"
      >
        2021-04-01T11:41:20.000Z
      </div>
    </td>
    <td>
      <div
        class="TableCell__TableCellInnerWrapper-sc-12y9mmp-0 dbZfon justify-end sm:border-l border-theme-secondary-300"
      >
        <span
          class="sm:hidden lg:table-cell"
        >
          1800000000.00000000 DARK
        </span>
        <span
          class="hidden lg:hidden sm:table-cell"
        >
          1800000000.00 DARK
        </span>
      </div>
    </td>
    <td
      class="hidden lg:table-cell"
    >
      <div
        class="TableCell__TableCellInnerWrapper-sc-12y9mmp-0 dbZfon justify-end sm:border-l border-theme-secondary-300"
      >
        18.00000000 DARK
      </div>
    </td>
  </tr>
</div>
`;
